tensorpipe/channel/cuda_basic/channel.cc

```diff
@@ -0,0 +1,358 @@
+      id_(std::move(id)) {}
+
+void Channel::Impl::init() {
+  loop_.deferToLoop([this]() { initFromLoop(); });
+}
+
+void Channel::Impl::initFromLoop() {
+  TP_DCHECK(loop_.inLoop());
+  closingReceiver_.activate(*this);
+}
+
+void Channel::send(
+    CudaBuffer buffer,
+    TDescriptorCallback descriptorCallback,
+    TSendCallback callback) {
+  impl_->send(buffer, std::move(descriptorCallback), std::move(callback));
+}
+
+void Channel::Impl::send(
+    CudaBuffer buffer,
+    TDescriptorCallback descriptorCallback,
+    TSendCallback callback) {
+  loop_.deferToLoop([this,
+                     buffer,
+                     descriptorCallback{std::move(descriptorCallback)},
+                     callback{std::move(callback)}]() mutable {
+    sendFromLoop(buffer, std::move(descriptorCallback), std::move(callback));
+  });
+}
+
+// Send memory region to peer.
+void Channel::Impl::sendFromLoop(
+    CudaBuffer buffer,
+    TDescriptorCallback descriptorCallback,
+    TSendCallback callback) {
+  TP_DCHECK(loop_.inLoop());
+
+  const uint64_t sequenceNumber = nextTensorBeingSent_++;
+  TP_VLOG(4) << "Channel " << id_ << " received a send request (#"
+             << sequenceNumber << ")";
+
+  if (error_) {
+    descriptorCallback(error_, std::string());
+    callback(error_);
+    return;
+  }
+
+  TP_VLOG(4) << "Channel " << id_
+             << " is copying buffer from CUDA device to CPU";
+  auto tmpBuffer = makeCudaPinnedBuffer(buffer.length);
+  TP_CUDA_CHECK(cudaMemcpyAsync(
+      tmpBuffer.get(),
+      buffer.ptr,
+      buffer.length,
+      cudaMemcpyDeviceToHost,
+      buffer.stream));
+
+  auto cudaCallback =
+      [this,
+       buffer,
+       tmpBuffer{std::move(tmpBuffer)},
+       descriptorCallback{std::move(descriptorCallback)},
+       callback{std::move(callback)}](cudaError_t cudaError) mutable {
+        TP_VLOG(4) << "Channel " << id_
+                   << " is done copying buffer from CUDA device to CPU";
+        TP_CUDA_CHECK(cudaError);
+
+        // TODO: Use eagerCallbackWrapper to ensure channel remains alive until
+        // this callback fires?
+
+        loop_.deferToLoop([this,
+                           buffer,
+                           tmpBuffer{std::move(tmpBuffer)},
+                           descriptorCallback{std::move(descriptorCallback)},
+                           callback{std::move(callback)}]() mutable {
+          if (error_) {
+            descriptorCallback(error_, std::string());
+            callback(error_);
+            return;
+          }
+
+          CpuBuffer cpuBuffer{tmpBuffer.get(), buffer.length};
+          // Keep tmpBuffer alive until cpuChannel_ is done sending it over.
+          callback = eagerCallbackWrapper_(
+              [tmpBuffer{std::move(tmpBuffer)}, callback{std::move(callback)}](
+                  Impl& impl) { callback(impl.error_); });
+          cpuChannel_->send(
+              cpuBuffer, std::move(descriptorCallback), std::move(callback));
```

Comment:
Would be nice if this could be a separate method, since it's rather big to be a lambda.

To reuse the approach we're using in the pipe, it would be nice if we could have a method for each "reaction", i.e., each thing that the pipe/channel needs to do when some "signal" from outside comes in. This will also make it easier to convert this to an explicit state machine if we so wish.